ci(autofix): move agent prompts into a project skill (#6306)
QwenLM/qwen-code: .qwen/skills/autofix/SKILL.md

   the relevant code.
2. Reject work that cannot be reproduced in headless Linux CI, including
   Windows/macOS-only bugs, real OAuth flows, IDE extension behavior, or human
   visual judgment.
3. Reject likely fixes requiring more than a small scoped change, architectural
   redesign, or product decisions.
4. If a report mixes symptoms, judge the reporter's primary complaint. If only
   a side symptom is fixable, skip this issue and mention the side symptom so a
   human can split it out. Do not mark that skip permanent solely for the side
   symptom.

Pick at most one issue: highest confidence first, forced tier-0 before tier-1,
then most recent among comparable candidates. It is valid to pick none.

Write `<workdir>/decision.json` with exactly this shape:

```json
{
  "go": 1234,
  "reason": "one paragraph: why this issue, suspected root cause, fix sketch, verification plan",
  "skip": [{ "number": 5678, "reason": "short reason", "permanent": false }]
}
```

Use `"go": null` when picking none. `"permanent": true` is only for issues that
are structurally unfixable by this bot and should never be rescanned.

## Mode: develop-issue

Use when invoked as `develop-issue`.

Inputs:

> 

Author: Leaving this as follow-up. Reusing or recovering an existing remote autofix branch is a broader retry policy change; this PR keeps the additive-commit rule and workflow ownership unchanged.

<!-- resolve-pr-comments -->

> 

Author: I am not adding automatic stale-branch recovery in this PR. If `origin/autofix/issue-<issue>` already exists, the additive push/create-PR path should stop for operator cleanup rather than overwrite existing remote history.

<!-- resolve-pr-comments -->

## Mode: address-review

Use when invoked as `address-review`.

Inputs:

- `--pr`: autofix pull request number.
- `--issue`: issue number fixed by the pull request.
- `--workdir`: directory containing `feedback.md`.
- `--conflict`: whether the branch conflicts with the base branch. Default:
  `false`.
- `--base`: base branch name. Default: `main`.

The workflow has already checked out `autofix/issue-<issue>`. Stay on that
branch. Do not create a new branch. First read the existing diff with
`git diff origin/<base>...HEAD`, then read `<workdir>/feedback.md`.

Classify every feedback point:

- Critical or merge-blocking: correctness bug, broken build/test, security
  problem, or a `CHANGES_REQUESTED` item that names a real defect. Verify it
  against current code before changing anything, then fix it minimally.
- Suggestion, nit, or optional hardening: use engineering judgment. Implement
  only suggestions that are reasonable, valuable, and within the PR scope. Skip
  low-value or over-engineered suggestions and explain why.

If `--conflict true`, merge `origin/<base>`, resolve every conflict by
understanding both sides, and include conflict notes in the summary. If false,

> do not merge unnecessarily.

Author: No code change here: the workflow invokes address-review with `--base main`, and the test now asserts that exact invocation. Making the mode less parameterized is cleanup beyond this prompt-migration PR.

<!-- resolve-pr-comments -->

> do not merge unnecessarily.

Author: No code change here: this workflow invokes the skill with `--base main`, and the prepare step also checks conflicts against `origin/main`. Supporting arbitrary base branches would be a broader workflow behavior change outside this refactor.

<!-- resolve-pr-comments -->

  drive-by refactors.
- Do not run project code, tests, builds, package scripts, or the CLI yourself.
  The workflow verification gate runs trusted checks after you exit.
- If confidence drops, stop cleanly and write the required failure file instead
  of making a speculative change.

## Mode: assess-candidates

Use when invoked as `assess-candidates`.

Inputs:

- `--workdir`: directory containing `candidates.json`. Default:
  `/tmp/autofix`.
- Candidate issues are in `<workdir>/candidates.json`.
- `autofixTier: 0` means manual dispatch or issue label event; treat it as
  highest priority.
- `autofixTier: 1` means a maintainer approved the issue for autonomous fixing
  with `autofix/approved`.

For each candidate, judge whether it is a reasonable, actionable bugfix or small
feature that an autonomous agent can confidently implement and verify:

1. Confirm the report is coherent and plausible in this codebase by locating
   the relevant code.
2. Reject work that cannot be reproduced in headless Linux CI, including
   Windows/macOS-only bugs, real OAuth flows, IDE extension behavior, or human
   visual judgment.

> 3. Reject likely fixes requiring more than a small scoped change, architectural

Reviewer: **[Suggestion]** Lost quantitative scope boundary. The old inline assess-candidates prompt required the likely fix to be "roughly <300 lines." This was replaced with the vague "more than a small scoped change." Without the numeric guardrail, the agent is more likely to accept issues whose fix requires hundreds of lines, increasing the risk of low-quality autonomous PRs.

```suggestion
3. Reject likely fixes requiring more than roughly 300 lines of change, architectural
   redesign, or product decisions.
```

_— qwen3.7-max via Qwen Code /review_

   redesign, or product decisions.
4. If a report mixes symptoms, judge the reporter's primary complaint. If only
   a side symptom is fixable, skip this issue and mention the side symptom so a
   human can split it out. Do not mark that skip permanent solely for the side
   symptom.

Pick at most one issue: highest confidence first, forced tier-0 before tier-1,
then most recent among comparable candidates. It is valid to pick none.

Write `<workdir>/decision.json` with exactly this shape:

```json
{
  "go": 1234,
  "reason": "one paragraph: why this issue, suspected root cause, fix sketch, verification plan",
  "skip": [{ "number": 5678, "reason": "short reason", "permanent": false }]
}
```

Use `"go": null` when picking none. `"permanent": true` is only for issues that
are structurally unfixable by this bot and should never be rescanned.

## Mode: develop-issue

Use when invoked as `develop-issue`.

Inputs:

- `--issue`: selected issue number.

> - `--workdir`: directory containing `candidates.json` and `decision.json`.

Reviewer: **[Suggestion]** `develop-issue` never explicitly instructs the agent to read `candidates.json` or `decision.json`. The old inline prompt stated: "Its full text is in /tmp/autofix/candidates.json and the assessment that selected it is in /tmp/autofix/decision.json." The new skill lists them as "Inputs" but never tells the agent to read them. By contrast, `address-review` (line 137) explicitly says "then read `<workdir>/feedback.md`." The assessment's `reason` field contains the fix sketch and verification plan — losing it means the agent works without context.

```suggestion
Implement the selected issue end to end in the checked-out repository:

Read `<workdir>/candidates.json` for the full issue text and `<workdir>/decision.json`
for the assessment that selected it.

Follow the project conventions in `AGENTS.md`…
```

_— qwen3.7-max via Qwen Code /review_

   missing capability or current gap. If you cannot establish the baseline,
   write `<workdir>/failure.md` and exit without committing.
3. Implement the minimal root-cause change. Avoid unrelated cleanup.
4. Add or update collocated Vitest tests expected to fail before the fix and
   pass after.
5. Describe the focused checks the workflow should run after you exit.
6. Re-read the full diff as a skeptical reviewer and fix issues you would flag.
7. Create one Conventional Commit on the branch, for example
   `fix(core): <summary> (#<issue>)`.
8. Write all required outputs:
   - `<workdir>/e2e-report.md`: baseline evidence, after behavior, exact
     suggested verification commands, and relevant output excerpts.
   - `<workdir>/pr-title.txt` and `<workdir>/pr-body.md`: use the project
     `prepare-pr` skill for issue `<issue>`.

If the fix is beyond confident reach, write `<workdir>/failure.md` with what
you learned and exit without committing.

## Mode: address-review

Use when invoked as `address-review`.

Inputs:

- `--pr`: autofix pull request number.
- `--issue`: issue number fixed by the pull request.

> - `--workdir`: directory containing `feedback.md`.

Reviewer: **[Suggestion]** `address-review` drops the self-review step. `develop-issue` (line 110) preserves "Re-read the full diff as a skeptical reviewer and fix issues you would flag" but the equivalent step is absent from `address-review`. Review-response commits are especially prone to introducing new issues because the agent is making targeted edits under guidance that may itself be partially incorrect — making this guard more important here, not less.

```suggestion
- Made a change: before committing, re-read the full diff as a skeptical reviewer and
  fix anything you would flag. Commit one Conventional Commit such as
  `fix(core): address review feedback (#<issue>)`, then write…
```

_— qwen3.7-max via Qwen Code /review_

- `--conflict`: whether the branch conflicts with the base branch. Default:
  `false`.
- `--base`: base branch name. Default: `main`.

The workflow has already checked out `autofix/issue-<issue>`. Stay on that
branch. Do not create a new branch. First read the existing diff with
`git diff origin/<base>...HEAD`, then read `<workdir>/feedback.md`.
Follow `AGENTS.md`. Keep collocated Vitest tests green; add or update tests
when feedback exposes a test gap.

Classify every feedback point:

- Critical or merge-blocking: correctness bug, broken build/test, security
  problem, or a `CHANGES_REQUESTED` item that names a real defect. Verify it
  against current code before changing anything, then fix it minimally.

> - Suggestion, nit, or optional hardening: use engineering judgment. Implement

Reviewer: **[Suggestion]** `address-review` drops the anti-scope-creep directive. The old inline prompt said "Prefer NOT to deviate from this PR's original direction and scope. Implement ONLY suggestions that are reasonable and genuinely valuable." The new skill says suggestions should be "within the PR scope" but lacks the strong "Prefer NOT to deviate" framing. Without it, the agent is more likely to implement tangential suggestions that expand PR scope.

```suggestion
- Suggestion, nit, or optional hardening: use engineering judgment. Prefer NOT to
  deviate from this PR's original direction and scope. Implement only suggestions
  that are reasonable, valuable, and within the PR scope. Skip low-value or
  over-engineered suggestions and explain why.
```

_— qwen3.7-max via Qwen Code /review_
